tools/protoprint/protoprint.py

```diff
@@ -726,20 +747,22 @@ def visit_file(self, file_proto, type_context, services, msgs, enums):
         return clang_format(header + formatted_services + formatted_enums + formatted_msgs)
 
 
-if __name__ == '__main__':
-    proto_desc_path = sys.argv[1]
+class ProtoprintTraverser:
+
+    def traverse_file(self, file_proto, visitor):
+        # This breaks the Visitor contract, ie class props should not be set on individual visitation.
+        visitor._frozen_proto = (
+            file_proto.options.Extensions[status_pb2.file_status].package_version_status ==
+            status_pb2.FROZEN)
+        return traverse.traverse_file(file_proto, visitor)
 
+
+def main(data=None):
     utils.load_protos()
 
-    file_proto = descriptor_pb2.FileDescriptorProto()
-    input_text = pathlib.Path(proto_desc_path).read_text()
-    if not input_text:
-        sys.exit(0)
-    text_format.Merge(input_text, file_proto)
-    dst_path = pathlib.Path(sys.argv[2])
-    utils.load_type_db(sys.argv[3])
-    api_version_file_path = pathlib.Path(sys.argv[4])
-    frozen_proto = file_proto.options.Extensions[
-        status_pb2.file_status].package_version_status == status_pb2.FROZEN
-    dst_path.write_bytes(
-        traverse.traverse_file(file_proto, ProtoFormatVisitor(api_version_file_path, frozen_proto)))
+    plugin.plugin([plugin.direct_output_descriptor('.proto', ProtoFormatVisitor, want_params=True)],
```

Comment:
@htuch i tried to separate this PR into test changes and proto_format check changes but it wasnt really possible because of this change here

basically, as soon as you switch `protoprint` to being a proper protoc plugin like the others then both the test and the proto_format check have to be updated to work with it

Reply:
Got it, that's fine.